chef/cookbooks/cinder/recipes/ceph.rb

```diff
@@ -17,36 +17,58 @@
 # Recipe:: ceph
 #
 
-glance_servers = search(:node, "roles:glance-server")
-if glance_servers.length > 0
-  glance_pool = glance_servers[0][:glance][:rbd][:store_pool]
-else   
-  glance_pool = nil
-end
+ceph_clients = {}
+ceph_keyrings = {}
 
 node[:cinder][:volumes].each_with_index do |volume, volid|
-  if volume['backend_driver'] == "rbd"
+  if volume[:backend_driver] == "rbd"
+
+    ceph_conf = volume[:rbd][:config_file]
+    admin_keyring = volume[:rbd][:admin_keyring]
 
-    check_ceph = Mixlib::ShellOut.new("ceph -k #{volume['rbd']['admin_keyring']} -c #{volume['rbd']['config_file']} -s | grep -q -e 'HEALTH_[OK|WARN]'")
+    check_ceph = Mixlib::ShellOut.new("ceph -k #{admin_keyring} -c #{ceph_conf} -s | grep -q -e 'HEALTH_[OK|WARN]'")
     check_ceph.run_command
 
     if check_ceph.exitstatus == 0
 
-      backend_id = "backend-#{volume['backend_driver']}-#{volid}"
+      backend_id = "backend-#{volume[:backend_driver]}-#{volid}"
+
       cinder_user = volume[:rbd][:user]
       cinder_pool = volume[:rbd][:pool]
+      ceph_clients[ceph_conf] = {} unless ceph_clients[ceph_conf]
+      ceph_keyrings[ceph_conf] = admin_keyring unless ceph_keyrings[ceph_conf]
+
+      cinder_pools = (ceph_clients[ceph_conf][cinder_user] || []) << cinder_pool
+      ceph_clients[ceph_conf][cinder_user] = cinder_pools
+
+      ceph_pool cinder_pool do
+        ceph_conf ceph_conf
+        admin_keyring admin_keyring
+      end
+    end
+  end
+end
+
+unless ceph_clients.empty?
+  glance_servers = search(:node, "roles:glance-server")
+  if glance_servers.length > 0
+    glance_pool = glance_servers[0][:glance][:rbd][:store_pool]
+  else
+    glance_pool = nil
+  end
 
-      cinder_pools = []
-      cinder_pools << cinder_pool
+  ceph_clients.each do |ceph_inst|
+    ceph_conf = ceph_inst[0]
+
+    ceph_inst[1].each_pair do |cinder_user, cinder_pools|
       cinder_pools << glance_pool unless glance_pool.nil?
+  
       allow_pools = cinder_pools.map{|p| "allow rwx pool=#{p}"}.join(", ")
       ceph_caps = { 'mon' => 'allow r', 'osd' => "allow class-read object_prefix rbd_children, #{allow_pools}" }
 
       ceph_client cinder_user do
-        unless volume['rbd']['use_crowbar']
-          ceph_conf  volume['rbd']['config_file']
-          admin_keyring  volume['rbd']['admin_keyring']
-        end
+        ceph_conf  ceph_conf
+        admin_keyring  ceph_keyrings[ceph_conf]
         caps ceph_caps
         keyname "client.#{cinder_user}"
         filename "/etc/ceph/ceph.client.#{cinder_user}.keyring"
```

Comment:
My proposal is to create user as `md5sum` hash function from `ceph_conf` and `cinder_user` string